src/target/riscv/riscv.c

```diff
@@ -4376,96 +4387,163 @@ static bool gdb_regno_cacheable(enum gdb_regno regno, bool is_write)
+			/* write_through */ false);
+}
+
+/**
+ * This function is used to change the value of a register. The new value may
+ * be cached, but it will be written to hart immediately.
+ */
+int riscv_write_register(struct target *target, enum gdb_regno regid,
+		riscv_reg_t value)
+{
+	return riscv_set_or_write_register(target, regid, value,
+			/* write_through */ true);
+}
+
+/**
+ * This function is used to get the value of a register. If possible, the value
+ * in cache will be updated.
+ */
 int riscv_get_register(struct target *target, riscv_reg_t *value,
 		enum gdb_regno regid)
 {
 	RISCV_INFO(r);
+	assert(r->get_register);
 
 	keep_alive();
 
-	struct reg *reg = &target->reg_cache->reg_list[regid];
+	if (!target->reg_cache) {
+		assert(!target_was_examined(target));
+		LOG_TARGET_DEBUG(target, "No cache, reading %s from target",
+				gdb_regno_name(regid));
+		return r->get_register(target, value, regid);
+	}
+
+	struct reg *reg = get_reg_cache_entry(target, regid);
 	if (!reg->exist) {
-		LOG_DEBUG("[%s] %s does not exist.",
-				  target_name(target), gdb_regno_name(regid));
+		LOG_TARGET_DEBUG(target, "Register %s does not exist.", reg->name);
 		return ERROR_FAIL;
 	}
 
-	if (reg && reg->valid) {
+	if (reg->valid) {
 		*value = buf_get_u64(reg->value, 0, reg->size);
-		LOG_DEBUG("[%s] %s: %" PRIx64 " (cached)", target_name(target),
-				  gdb_regno_name(regid), *value);
+		LOG_TARGET_DEBUG(target, "Read %s: 0x%" PRIx64 " (cached)", reg->name,
+				*value);
 		return ERROR_OK;
 	}
 
-	/* TODO: Hack to deal with gdb that thinks these registers still exist. */
-	if (regid > GDB_REGNO_XPR15 && regid <= GDB_REGNO_XPR31 &&
-			riscv_supports_extension(target, 'E')) {
-		*value = 0;
-		return ERROR_OK;
-	}
+	LOG_TARGET_DEBUG(target, "Reading %s from target", reg->name);
+	if (r->get_register(target, value, regid) != ERROR_OK)
+		return ERROR_FAIL;
 
-	int result = r->get_register(target, value, regid);
+	buf_set_u64(reg->value, 0, reg->size, *value);
+	reg->valid = gdb_regno_cacheable(regid, /* is write? */ false);
+	reg->dirty = false;
 
-	if (result == ERROR_OK) {
-		/* Update the cache in case we're called from
-		 * riscv_save_register(). */
-		buf_set_u64(reg->value, 0, reg->size, *value);
-		reg->valid = gdb_regno_cacheable(regid, false);
-	}
-
-	LOG_DEBUG("[%s] %s: %" PRIx64, target_name(target),
-			gdb_regno_name(regid), *value);
-	return result;
+	LOG_TARGET_DEBUG(target, "Read %s: 0x%" PRIx64, reg->name, *value);
+	return ERROR_OK;
 }
 
+/**
+ * This function is used to save the value of a register in cache. The register
+ * is marked as dirty, and writeback is delayed for as long as possible.
+ */
 int riscv_save_register(struct target *target, enum gdb_regno regid)
 {
+	assert(target->state == TARGET_HALTED &&
+			"Doesn't make sense to populate register cache on non-halted targets.");
+	assert(gdb_regno_cacheable(regid, /* is write? */ false) &&
+			"Only cacheable registers can be saved.");
```

Comment:
I am not aware that this "hack" to [include the textual message into the assert](https://stackoverflow.com/a/3692961) would be used anywhere else in OpenOCD.

For that reason, and to make the future upstreaming more smooth, I'd recommend to not introduce this new pattern.

I think it is sufficient to have the usual assert condition + the explanation in a comment.

```suggestion
	/* Doesn't make sense to populate register cache on non-halted targets. */
	assert(target->state == TARGET_HALTED);
	/* Only cacheable registers can be saved. */
	assert(gdb_regno_cacheable(regid, /* is write? */ false));
```

Reply:
FWIW, I'm fine with this. If upstream complains then we can easily translate them back. This style seems better to me because when the user sees the assert at least there's some indication on what went wrong.

Reply:
I'm not happy to see this creative abuse of `&&` inside the asserts, but granted, the error message that gets displayed is added value. And this pattern doesn't seem to be totally uncommon (there e.g. several uses in the jimtcl library, used by OpenOCD). So please feel free to leave it there, and we will see later whether there will be any push-back from the upstream.

As a related note: Most of OpenOCD code apparently follows a different style: `LOG_ERROR("BUG: ...")` is used in places where common `assert`s would seem to be more fitting.